system/lib/wasmfs/wasmfs.cpp

```diff
@@ -61,31 +62,103 @@ __wasi_errno_t __wasi_fd_write(__wasi_fd_t fd,
                                const __wasi_ciovec_t* iovs,
                                size_t iovs_len,
                                __wasi_size_t* nwritten) {
+  if (iovs_len < 0) {
+    return __WASI_ERRNO_INVAL;
+  }
+
   auto openFile = FileTable::get()[fd];
 
   if (!openFile) {
     return __WASI_ERRNO_BADF;
   }
 
-  auto file = openFile.locked().getFile()->dynCast<DataFile>();
+  auto lockedOpenFile = openFile.locked();
+  auto* file = lockedOpenFile.getFile()->dynCast<DataFile>();
 
   // If file is nullptr, then the file was not a DataFile.
+  // TODO: change to add support for symlinks.
   if (!file) {
     return __WASI_ERRNO_ISDIR;
   }
 
   auto lockedFile = file->locked();
 
-  __wasi_size_t num = 0;
+  off_t offset = lockedOpenFile.position();
   for (size_t i = 0; i < iovs_len; i++) {
     const uint8_t* buf = iovs[i].buf;
-    __wasi_size_t len = iovs[i].buf_len;
+    size_t len = iovs[i].buf_len;
+
+    // Check if the sum of the buf_len values overflows a uint64_t (file size).
```

Comment:
I think we realized during the meeting that the maximum file size is 63 bits, so the comment might be misleading.

In fact, given that, is checking `uint64_t` the right thing? Correct me if I'm wrong, but I think we can use `off_t` everywhere instead of `uint64_t`. And also we don't need to add `wasmfs_offset_t`.

But maybe I missed something in the meeting?

Reply:
I think I had `uint64_t` here because overflow on signed types is undefined. However, the new function to check for overflow will explicitly verify this. 

You're right, I am pretty sure that we can use`off_t` here. Would I need to cast len to a `off_t` then I guess?

Reply:
Good point that checking for 64-bit overflow is not the same as checking for 63-bit overflow, @kripken! I guess it will be important to have a carefully-vetted helper for checking overflow. @ethanalee `len` will be promoted to an `off_t` during the addition, but you might want to cast it explicitly when passing it to the overflow check function.